book-content/chapters/10-deriving-types.md

```diff
@@ -0,0 +1,1460 @@
+
+// ---cut---
+type Group = ProgramModeMap["GROUP"];
+//   ^?
+```
+
+With this change, the `Group` type will be in sync with the `ProgramModeEnumMap`'s `group` value. This means our test will pass as expected.
+
+### Solution 4: Unions with Indexed Access Types
+
+In order to create the `PlannedPrograms` type, we can use an indexed access type to extract a union of the `ProgramModeMap` values that include "`planned`":
+
+```typescript
+type Key = "PLANNED_ONE_ON_ONE" | "PLANNED_SELF_DIRECTED";
+type PlannedPrograms = ProgramModeMap[Key];
+```
+
+With this change, the `PlannedPrograms` type will be a union of `planned1on1` and `plannedSelfDirected`, which means our test will pass as expected.
+
+### Solution 5: Extract a Union of All Values
+
+Using `keyof` and `typeof` together is the solution to this problem.
+
+The most condensed solution looks like this:
+
+```typescript
+type AllPrograms = (typeof programModeEnumMap)[keyof typeof programModeEnumMap];
+```
+
+Using an intermediate type, you could first use `typeof programModeEnumMap` to create a type from the `programModeEnumMap` object, then use `keyof` to extract the keys:
+
+```ts twoslash
+export const programModeEnumMap = {
+  GROUP: "group",
+  ANNOUNCEMENT: "announcement",
+  ONE_ON_ONE: "1on1",
+  SELF_DIRECTED: "selfDirected",
+  PLANNED_ONE_ON_ONE: "planned1on1",
+  PLANNED_SELF_DIRECTED: "plannedSelfDirected",
+} as const;
+
+// ---cut---
+type ProgramModeMap = typeof programModeEnumMap;
+type AllPrograms = ProgramModeMap[keyof ProgramModeMap];
+//   ^?
+```
+
+Either solution results in a union of all values from the `programModeEnumMap` object, which means our test will pass as expected.
+
+### Solution 6: Create a Union from an `as const` Array
+
+When using `typeof` and `keyof` with indexed access type, we can extract all of the values, but we also get some unexpected values like a `6` and an `IterableIterator` function:
+
+```ts twoslash
+export const programModes = [
+  "group",
+  "announcement",
+  "1on1",
+  "selfDirected",
+  "planned1on1",
+  "plannedSelfDirected",
+] as const;
+
+// ---cut---
+type AllPrograms = (typeof programModes)[keyof typeof programModes];
+//   ^?
+```
+
+The additional stuff being extracted causes the test to fail because it is only expecting the original values instead of numbers and functions.
+
+Recall that we can access the first element using `programModes[0]`, the second element using `programModes[1]`, and so on. This means that we could use a union of all possible index values to extract the values from the `programModes` array:
+
+```typescript
+type AllPrograms = (typeof programModes)[0 | 1 | 2 | 3 | 4 | 5];
+```
+
+This solution makes the test pass, but it doesn't scale well. If the `programModes` array were to change, we would need to update the `AllPrograms` type manually.
+
+Instead, we can use the `number` type as the argument to the indexed access type to represent all possible index values:
```

Comment:
It's worth mentioning that it works like that because tuples are... arrays. We can have fixed-size tuples like:
```ts
type Tup = [string, number]
```
So one could imagine that we could only access this with `0` or `1` but to make tuples more practical they also have a **numeric index signature** that is a union of all possible property types here. And since a numeric index signature is there (just like on array types) it's permitted to use `number` to index into them.

This introduces unsoundness in scenarios like this:
```ts
function test(tup: [string, number], index: number) {
  const result = tup[index]
  //    ^?
}

test(['foo', 42], 5);
```

This could be seen as odd, the index signature isn't used when TS knows the exact number we are using for indexing... but as soon as we just introduce a little bit of indirection to lose the literal type it becomes fine with it:
```ts
const tup = ['foo', 42] as const
tup[5] // error

let num = 5
tup[num] // ok
```

It's a side-step from the type theory with the aim to make the language more approachable and useful and to allow for common patterns (even if they sometimes might bite back)